### ISSUE TITLE
Clarification: batch size definition (per-device vs global) & recommended LR for A100-80GB reproduction

### ISSUE THREAD

**Issue by misoyuri** [hqsiswiliam/hira#9]
Hi authors, thanks for the repo.

I’m trying to reproduce the paper results on A100 80GBs (Llama-3-8B, commonsense common_170k, r_ab=32).
But I’m confused about the batch size / LR settings because they appear different across the repo and the paper:
- README single-GPU example uses --lr=1e-3 --batch=16
- Paper Appendix (Table 7) lists Batch Size = 32 with LR ∈ {1e-4, 2e-4}
- Paper Appendix A.6 / Table 12 (memory & runtime analysis) uses Batch Size = 72

Could you please clarify:
1) In train_hira.py, does --batch mean per-device micro-batch or global/effective batch? If per-device, what grad_accum / #GPUs correspond to the reported batch values?
2) For 1×A100 80GB reproduction of the main commonsense results, what exact training setup should we use (LR, batch, grad_accum, bf16/fp16, and whether to use DeepSpeed/config)?
3) DoRA baseline question: In the official DoRA repo, the 1-GPU setting uses batch=16. For comparison with HiRA/LoRA in your paper, should we keep DoRA as-is, or match the same effective/global batch size (and other key settings)?

Thanks!

**Comment by hqsiswiliam** [comment]
Dear @misoyuri,

Thanks for your interest in our work.

1. In `train_hira.py`, `--batch` refers to the **per-device batch size** (i.e., micro-batch size).

2. For **single A100-80GB** experiments, we used learning rates of **1e-3 or 2e-4**, **no gradient accumulation**, **bf16**, and **no DeepSpeed**, as training was conducted on a single GPU. The different batch sizes reported in the paper (e.g., 32 or 72) correspond to different experimental settings and memory/runtime analyses rather than a single fixed configuration.

3. For the **DoRA baseline**, we recommend tuning it under its own optimal settings. Since DoRA has different learning dynamics compared to HiRA/LoRA, matching exact hyperparameters is not always necessary or optimal.

Hope this helps clarify the setup. Please feel free to follow up if anything is still unclear.

Best regards,
Qiushi

**Comment by misoyuri** [comment]
Thanks for the clarification, @hqsiswiliam !

For reproducibility and to better understand the comparison protocol in the paper, could you share your overall experimental setup policy?

1) Reproducibility: For each main table (and key appendix tables), what exact training/eval configs were used (LR, per-device batch, grad accumulation, precision bf16/fp1, seed list)? 
A table→command (or table→config) mapping would be perfect.

2) Comparison protocol: When comparing HiRA/LoRA/DoRA (and other PEFT baselines), did you:
   (a) match a common effective/global batch size and training budget across methods, or
   (b) tune each method under its own best/optimal settings (LR/batch/etc.)?
If (b), could you share the tuning ranges and the budget used for each baseline?

Thanks again!

**Comment by hqsiswiliam** [comment]
Hi @misoyuri,
Thanks for the follow-up questions. I’m happy to clarify our experimental setup and comparison protocol.
```
Hardware & batch size:
8×V100-32GB: batch size = 16 per device
1×A100-80GB: batch size = 72
Learning rate:
8×V100-32GB: 1e-3
1×A100-80GB: 2e-4
Warmup steps=5%–10%
Weight decay=0.01
```

Comparison protocol.
We follow strategy (b): each method is tuned under its own best-performing configuration. Since different PEFT methods have different computational characteristics, we tune hyperparameters independently to match their training dynamics. Concretely, for each method we search:
```
Learning rate: [1e-3, 1e-5]
Batch size: [16, 72]
Warmup steps=5%–10%
Weight decay: 0.01
```
The computational budget remains the same as HiRA's (1x A100-80G or 8x V100-32G).

By the way, for ChatGPT, LoRA, and DoRA baselines in Table 1, we directly report the results from the DoRA paper, as stated in the caption:
> Results for ChatGPT, LoRA, and DoRA are sourced from (Liu et al., 2024).

I hope these would be helpful for you.

**Comment by misoyuri** [comment]
Hi @hqsiswiliam , thank you so much for the detailed clarification.

This answers my questions clearly — especially the batch/LR settings per GPU and the protocol of tuning each method.
The note about Table 1 baselines being taken from the DoRA paper is also very helpful.

I’ll follow this setup for my reproduction, and I’ll follow up if I run into any unexpected issues.

Thanks again!

**Comment by misoyuri** [comment]
Hi @hqsiswiliam, thanks again for the previous detailed clarification — I followed the recipe you shared and tried to reproduce the results.

Unfortunately, I couldn’t reproduce the numbers well with that exact setup on my side. 
With a different setting, I can get *reasonable* performance, so I suspect this might be due to environment differences (e.g., hardware / CUDA / driver / PyTorch version affecting kernels, determinism, etc.). 
If you have any recommended “known-good” environment details (CUDA/PyTorch versions, bf16/fp16, etc.), I’d really appreciate it.

Also, I have a question about **MetaMath** training dataset preparation and answer-formatting:

* I downloaded the dataset from Hugging Face ([MetaMathQA](https://huggingface.co/datasets/meta-math/MetaMathQA))
  * Or [pissa-dataset](https://huggingface.co/datasets/fxmeng/pissa-dataset/tree/main/metamath) could be the candidate
 * I converted the keys to match your expected schema (using `sQuestion` and `lSolutions`) and use it as `train.json`.

However, during prediction, I noticed the model often **does not** end with the desired format like:
`the answer is {}`
(i.e., the final answer string is not reliably produced).

Could you share how you prepare the MetaMath training data and how you enforce the final-answer format?
In particular:

1. Do you apply any additional preprocessing/filtering on MetaMathQA?
2. Could you share meta math train.json which you used?

Any guidance or a snippet of the formatting template would help a lot.
Thank you!
